File: packages/runtime-go/core/validate.go
package core

import (
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

// Runtime response validation.
//
// A generated struct is a *claim* about what the server sends. Go makes checking it more necessary
// than elsewhere, not less: `encoding/json` **silently ignores** a field whose type does not match, so
// a server sending `"seats": "many"` for an `int64` leaves the field at zero and the caller cannot
// tell that from a legitimate zero. Nothing fails, and the wrong number propagates.
//
// That is why validation runs against the *decoded generic value*, before unmarshalling into the typed
// struct. Validating the struct afterwards could not work — by then the mismatch has already been
// discarded. It costs a second `json.Unmarshal` of the same buffer, which is the honest price of
// knowing.
//
// What is checked and what deliberately is not mirrors the other runtimes exactly (SPEC.md
// §3.4.1.1): required fields, declared types, and nullability; never unknown fields, never enum
// membership.

// SchemaKind identifies a descriptor's shape.
type SchemaKind string

const (
	SchemaString  SchemaKind = "str"
	SchemaNumber  SchemaKind = "num"
	SchemaInteger SchemaKind = "int"
	SchemaBool    SchemaKind = "bool"
	SchemaAny     SchemaKind = "any"
	SchemaArray   SchemaKind = "arr"
	SchemaObject  SchemaKind = "obj"
	SchemaMap     SchemaKind = "map"
	SchemaNull    SchemaKind = "null"
	SchemaOr      SchemaKind = "or"
	SchemaRef     SchemaKind = "ref"
)

// SchemaField is one declared property of an object descriptor.
type SchemaField struct {
	Name     string `json:"n"`
	Schema   Schema `json:"s"`
	Required bool   `json:"r,omitempty"`
}

// Schema is a response-shape descriptor.
//
// One struct with optional fields rather than an interface hierarchy: Go has no sum type, and a
// hierarchy would mean a type assertion at every step of the walk for no gain in safety. The `json`
// tags are short because the whole table is embedded as JSON in generated code.
type Schema struct {
	K SchemaKind `json:"k"`
	// I is the item schema for arr, and the inner schema for null.
	I *Schema `json:"i,omitempty"`
	// V is the value schema for map.
	V *Schema `json:"v,omitempty"`
	// F is the declared fields of obj.
	F []SchemaField `json:"f,omitempty"`
	// A is the additional-properties schema of obj, when the spec declared one.
	A *Schema `json:"a,omitempty"`
	// O is the branches of or.
	O []Schema `json:"o,omitempty"`
	// N is the table key for ref.
	N string `json:"n,omitempty"`
}

// SchemaTable maps a type name to its descriptor. Recursive types terminate through it.
type SchemaTable map[string]Schema

// ValidationMode is how strictly to enforce the declared response shape.
type ValidationMode string

const (
	// ValidationStrict returns a *ResponseValidationError. The default.
	ValidationStrict ValidationMode = "strict"
	// ValidationWarn reports to the client's Logf and returns the value.
	ValidationWarn ValidationMode = "warn"
	// ValidationOff skips the check.
	ValidationOff ValidationMode = "off"
)

// ValidationProblem is one contract violation.
type ValidationProblem struct {
	// Path is a JSON path from the response root, e.g. `data[0].email`.
	Path    string
	Message string
}

// ResponseValidationError is returned when a response does not match the declared shape.
//
// Deliberately not an *APIError: the request succeeded and the *contract* was violated. That is a
// different problem for the caller — and usually a different problem for the API owner — so
// `errors.As(err, &apiErr)` must not match it.
type ResponseValidationError struct {
	// Operation names the method, e.g. `Orgs.ListMembers`.
	Operation string
	Problems  []ValidationProblem
	// Body is the raw response, so a caller who wants to proceed anyway still has it.
	Body []byte
}

func (e *ResponseValidationError) Error() string {
	// The first few, not all: a response violating its contract fifty times over is one broken
	// contract, and a fifty-line message buries the useful part.
	limit := len(e.Problems)
	if limit > 3 {
		limit = 3
	}
	parts := make([]string, 0, limit)
	for _, problem := range e.Problems[:limit] {
		path := problem.Path
		if path == "" {
			path = "response"
		}
		parts = append(parts, path+" "+problem.Message)
	}
	tail := ""
	if rest := len(e.Problems) - limit; rest > 0 {
		tail = fmt.Sprintf(", and %d more", rest)
	}
	return fmt.Sprintf("%s: the response did not match the API's declared shape — %s%s",
		e.Operation, strings.Join(parts, "; "), tail)
}

func (e *ResponseValidationError) sdkError() {}

const (
	maxProblems = 50
	maxDepth    = 64
)

// Validate checks a decoded value against a schema, returning the problems found.
//
// An empty result means it conformed. Collecting rather than returning early lets warn mode report
// everything, and lets one error carry every violation.
func Validate(value any, schema Schema, table SchemaTable) []ValidationProblem {
	var problems []ValidationProblem
	walkSchema(value, schema, table, "", &problems, 0)
	return problems
}

func walkSchema(value any, schema Schema, table SchemaTable, path string, problems *[]ValidationProblem, depth int) {
	if len(*problems) >= maxProblems || depth > maxDepth {
		return
	}

	switch schema.K {
	case SchemaAny:
		return

	case SchemaRef:
		target, ok := table[schema.N]
		if !ok {
			// A dangling reference is a generator bug, not a server one. Reported as such rather than
			// failing the response, because punishing the user for our mistake is the wrong trade.
			appendProblem(problems, path, "references unknown schema `"+schema.N+"`")
			return
		}
		walkSchema(value, target, table, path, problems, depth+1)

	case SchemaNull:
		if value == nil {
			return
		}
		if schema.I != nil {
			walkSchema(value, *schema.I, table, path, problems, depth+1)
		}

	case SchemaString:
		if _, ok := value.(string); !ok {
			appendProblem(problems, path, mismatchMessage("a string", value))
		}

	case SchemaNumber:
		number, ok := value.(float64)
		if !ok || math.IsNaN(number) || math.IsInf(number, 0) {
			appendProblem(problems, path, mismatchMessage("a number", value))
		}

	case SchemaInteger:
		// JSON numbers decode to float64 through `any`, so integrality is a value check rather than a
		// type check.
		number, ok := value.(float64)
		if !ok || math.IsNaN(number) || math.IsInf(number, 0) || number != math.Trunc(number) {
			appendProblem(problems, path, mismatchMessage("an integer", value))
		}

	case SchemaBool:
		if _, ok := value.(bool); !ok {
			appendProblem(problems, path, mismatchMessage("a boolean", value))
		}

	case SchemaArray:
		items, ok := value.([]any)
		if !ok {
			appendProblem(problems, path, mismatchMessage("an array", value))
			return
		}
		if schema.I == nil {
			return
		}
		for index, item := range items {
			walkSchema(item, *schema.I, table, fmt.Sprintf("%s[%d]", path, index), problems, depth+1)
			if len(*problems) >= maxProblems {
				return
			}
		}

	case SchemaMap:
		object, ok := value.(map[string]any)
		if !ok {
			appendProblem(problems, path, mismatchMessage("an object", value))
			return
		}
		if schema.V == nil {
			return
		}
		// Sorted so a failure is reproducible; Go's map iteration order is deliberately random.
		for _, key := range sortedMapKeys(object) {
			walkSchema(object[key], *schema.V, table, joinPath(path, key), problems, depth+1)
			if len(*problems) >= maxProblems {
				return
			}
		}

	case SchemaObject:
		object, ok := value.(map[string]any)
		if !ok {
			appendProblem(problems, path, mismatchMessage("an object", value))
			return
		}
		declared := make(map[string]bool, len(schema.F))
		for _, field := range schema.F {
			declared[field.Name] = true
			inner, present := object[field.Name]
			if !present {
				if field.Required {
					appendProblem(problems, joinPath(path, field.Name), "is required but was absent")
				}
				continue
			}
			// A present-but-null value is walked rather than skipped, so a `null` in a non-nullable
			// field is reported as the type mismatch it is.
			walkSchema(inner, field.Schema, table, joinPath(path, field.Name), problems, depth+1)
			if len(*problems) >= maxProblems {
				return
			}
		}
		// Unknown fields are never a problem. A server adding a field must not break a client, and
		// this is where that promise is kept.
		if schema.A != nil {
			for _, key := range sortedMapKeys(object) {
				if declared[key] {
					continue
				}
				walkSchema(object[key], *schema.A, table, joinPath(path, key), problems, depth+1)
				if len(*problems) >= maxProblems {
					return
				}
			}
		}

	case SchemaOr:
		// A union passes if any branch does. Only the *closest* failure is reported when none do:
		// listing every branch's complaints for a three-way union is noise, and the branch that got
		// furthest is almost always the one the server meant.
		var best []ValidationProblem
		for _, branch := range schema.O {
			var branchProblems []ValidationProblem
			walkSchema(value, branch, table, path, &branchProblems, depth+1)
			if len(branchProblems) == 0 {
				return
			}
			if best == nil || len(branchProblems) < len(best) {
				best = branchProblems
			}
		}
		if best == nil {
			appendProblem(problems, path, "matched no variant of the union")
			return
		}
		*problems = append(*problems, best...)
	}
}

func appendProblem(problems *[]ValidationProblem, path, message string) {
	if len(*problems) >= maxProblems {
		return
	}
	*problems = append(*problems, ValidationProblem{Path: path, Message: message})
}

func mismatchMessage(expected string, actual any) string {
	return "should be " + expected + " but was " + describeValue(actual)
}

// describeValue names a value's shape and never its content: the value may be a secret — a token
// echoed back, a key in an error body — and an error message ends up in logs.
func describeValue(value any) string {
	switch typed := value.(type) {
	case nil:
		return "null"
	case string:
		return "a string"
	case bool:
		return "a boolean"
	case float64:
		if typed == math.Trunc(typed) && !math.IsInf(typed, 0) {
			return "an integer"
		}
		return "a number"
	case []any:
		return "an array"
	case map[string]any:
		return "an object"
	default:
		return fmt.Sprintf("%T", value)
	}
}

func joinPath(path, key string) string {
	safe := key != ""
	for index, r := range key {
		isLetter := r == '_' || (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z')
		isDigit := r >= '0' && r <= '9'
		if !isLetter && !(isDigit && index > 0) {
			safe = false
			break
		}
	}
	if path == "" {
		if safe {
			return key
		}
		return "[" + strconv.Quote(key) + "]"
	}
	if safe {
		return path + "." + key
	}
	return path + "[" + strconv.Quote(key) + "]"
}

func sortedMapKeys(object map[string]any) []string {
	keys := make([]string, 0, len(object))
	for key := range object {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

// MustParseSchemaTable decodes an embedded descriptor table.
//
// Generated code embeds the table as compact JSON and parses it once at package initialisation.
// Go struct literals for the same data would be several times larger — Stripe's table is a few
// thousand descriptors — and slower to compile, so the JSON blob is both smaller in the repository and
// cheaper in the build. Parsing costs a few milliseconds once.
//
// Panics on malformed input, deliberately: the input is generated by besdk and embedded in the same
// binary, so a failure here is a generator bug that should surface at start-up rather than on the
// first request.
func MustParseSchemaTable(encoded string) SchemaTable {
	var table SchemaTable
	if err := json.Unmarshal([]byte(encoded), &table); err != nil {
		panic(fmt.Sprintf("sdk: embedded schema table is malformed: %v", err))
	}
	return table
}

// AsSlice reads a decoded value as a slice, returning nil when it is not one.
//
// Used by generated pagination code to validate items before they are unmarshalled into a typed slice.
// Tolerant rather than strict: a malformed envelope is already the decoder's problem to report, and
// panicking inside a validation helper would turn a server's bad response into a crash.
func AsSlice(value any) []any {
	if items, ok := value.([]any); ok {
		return items
	}
	return nil
}

// PrefixPaths rewrites problem paths to include the index of the item they came from.
//
// A paginator has already unwrapped the envelope, so a problem reported as `email` needs to become
// `[0].email` for the caller to know which item it was.
func PrefixPaths(problems []ValidationProblem, index int) []ValidationProblem {
	out := make([]ValidationProblem, len(problems))
	for i, problem := range problems {
		path := problem.Path
		if path == "" {
			path = fmt.Sprintf("[%d]", index)
		} else if strings.HasPrefix(path, "[") {
			path = fmt.Sprintf("[%d]%s", index, path)
		} else {
			path = fmt.Sprintf("[%d].%s", index, path)
		}
		out[i] = ValidationProblem{Path: path, Message: problem.Message}
	}
	return out
}

File: packages/runtime-typescript/src/validate.ts
/**
 * Runtime response validation.
 *
 * A generated type is a *claim* about what the server sends. Without checking it, the claim is
 * unverified — `widget.name.trim()` on a response missing `name` fails inside the caller's code with
 * `Cannot read properties of undefined`, and nothing points at the API that broke its contract.
 *
 * The generated SDK emits a compact descriptor table; this file is the one reviewed walker over it.
 * That split is deliberate (SPEC.md §3.4.1.1): a validator library would be a runtime dependency in a
 * package whose selling point is having none, and generated validator *code* would scale output size
 * with the type count — Stripe has 1,440 types.
 *
 * What is checked: required fields, declared types, and nullability. What is deliberately not:
 * unknown fields (a server adding one must never break a client) and enum membership (a server adding
 * a value must not become a decode failure — that is the whole reason open enums exist).
 */

import { SDKError } from './errors.js';

/**
 * A schema descriptor.
 *
 * Single-character keys because this table is emitted once per type and some specs have thousands.
 * The readability that matters is this file's, and this file is hand-written.
 */
export type Schema =
  /** string */
  | { readonly k: 'str' }
  /**
   * An RFC 3339 timestamp, revived as a `Date`.
   *
   * Only `date-time`. A `format: date` stays `str`, because JavaScript has no date-only type and
   * `new Date('2026-08-06')` is midnight *UTC* — so a caller in UTC-5 reading `.getDate()` gets the
   * day before. Handing someone a `Date` for a calendar date is how they walk into that
   * (SPEC.md §3.4.1.2).
   */
  | { readonly k: 'date' }
  /** number */
  | { readonly k: 'num' }
  /** integer — a number with no fractional part */
  | { readonly k: 'int' }
  /** boolean */
  | { readonly k: 'bool' }
  /** anything, including absent. `schema: {}` in the spec, or a union we chose not to narrow. */
  | { readonly k: 'any' }
  /** array */
  | { readonly k: 'arr'; readonly i: Schema }
  /** object with declared fields; extra keys are always allowed */
  | {
      readonly k: 'obj';
      readonly f: readonly (readonly [name: string, schema: Schema, required?: 1])[];
      /** Value schema for additional properties, when the spec declared one. */
      readonly a?: Schema;
    }
  /** map with homogeneous values */
  | { readonly k: 'map'; readonly v: Schema }
  /** the value may be null */
  | { readonly k: 'null'; readonly i: Schema }
  /** one of several shapes; passes if any branch passes */
  | { readonly k: 'or'; readonly o: readonly Schema[] }
  /** reference into the descriptor table, which is how recursive types terminate */
  | { readonly k: 'ref'; readonly n: string };

/** The emitted descriptor table: type name to schema. */
export type SchemaTable = Readonly<Record<string, Schema>>;

/** How strictly to enforce the contract. */
export type ValidationMode = 'strict' | 'warn' | 'off';

/** One contract violation. */
export interface ValidationProblem {
  /** JSON path from the response root, e.g. `data[0].email`. */
  readonly path: string;
  readonly message: string;
}

/**
 * Thrown when a response does not match the shape the spec promised.
 *
 * Deliberately **not** an `APIError`: the request succeeded and the *contract* was violated. That is a
 * different problem for the caller — and usually a different problem for the API owner — so it must
 * not be swallowed by a `catch` that meant "handle a 4xx".
 */
export class ResponseValidationError extends SDKError {
  readonly problems: readonly ValidationProblem[];
  /** The decoded body, so a caller who wants to proceed anyway still has it. */
  readonly body: unknown;

  constructor(operation: string, problems: readonly ValidationProblem[], body: unknown) {
    super(ResponseValidationError.describe(operation, problems));
    this.name = 'ResponseValidationError';
    this.problems = problems;
    this.body = body;
  }

  private static describe(operation: string, problems: readonly ValidationProblem[]): string {
    // The first few, not all of them: a response with a hundred violations is one broken contract,
    // and a hundred-line message buries the useful part.
    const shown = problems.slice(0, 3).map((p) => `${p.path || 'response'} ${p.message}`);
    const rest = problems.length - shown.length;
    const tail = rest > 0 ? `, and ${rest} more` : '';
    return `${operation}: the response did not match the API's declared shape — ${shown.join('; ')}${tail}`;
  }
}

const MAX_PROBLEMS = 50;

/**
 * Check a decoded body against a schema.
 *
 * Returns the problems found; an empty array means it conformed. Collecting rather than throwing lets
 * `warn` mode report without failing, and lets the error carry every violation at once.
 *
 * Stops at {@link MAX_PROBLEMS}. A response that violates its contract fifty times over is one broken
 * contract, and walking a large malformed payload to completion costs more than the information is
 * worth.
 */
export function validate(value: unknown, schema: Schema, table: SchemaTable): ValidationProblem[] {
  const problems: ValidationProblem[] = [];
  walk(value, schema, table, '', problems, 0);
  return problems;
}

/** Guard against a descriptor table with a reference cycle that does not consume input. */
const MAX_DEPTH = 64;

function walk(
  value: unknown,
  schema: Schema,
  table: SchemaTable,
  path: string,
  problems: ValidationProblem[],
  depth: number,
): void {
  if (problems.length >= MAX_PROBLEMS || depth > MAX_DEPTH) return;

  switch (schema.k) {
    case 'any':
      return;

    case 'ref': {
      const target = table[schema.n];
      if (target === undefined) {
        // A dangling reference is a generator bug, not a server one. Reported as such rather than
        // failing the response, because punishing the user for our mistake is the wrong trade.
        problems.push({ path, message: `references unknown schema \`${schema.n}\`` });
        return;
      }
      walk(value, target, table, path, problems, depth + 1);
      return;
    }

    case 'null':
      if (value === null) return;
      walk(value, schema.i, table, path, problems, depth + 1);
      return;

    case 'str':
      if (typeof value !== 'string') problems.push(mismatch(path, 'a string', value));
      return;

    case 'date':
      // Validated as a string *and* as parseable. A field the spec declared a timestamp but which
      // holds `"soon"` is a contract violation, and reporting it here is more useful than handing the
      // caller an `Invalid Date` to discover later.
      if (typeof value !== 'string') {
        problems.push(mismatch(path, 'a timestamp string', value));
      } else if (Number.isNaN(Date.parse(value))) {
        problems.push({ path, message: 'should be an RFC 3339 timestamp but could not be parsed' });
      }
      return;

    case 'num':
      if (typeof value !== 'number' || !Number.isFinite(value)) {
        problems.push(mismatch(path, 'a number', value));
      }
      return;

    case 'int':
      if (typeof value !== 'number' || !Number.isInteger(value)) {
        problems.push(mismatch(path, 'an integer', value));
      }
      return;

    case 'bool':
      if (typeof value !== 'boolean') problems.push(mismatch(path, 'a boolean', value));
      return;

    case 'arr': {
      if (!Array.isArray(value)) {
        problems.push(mismatch(path, 'an array', value));
        return;
      }
      for (let index = 0; index < value.length; index++) {
        walk(value[index], schema.i, table, `${path}[${index}]`, problems, depth + 1);
        if (problems.length >= MAX_PROBLEMS) return;
      }
      return;
    }

    case 'map': {
      if (!isRecord(value)) {
        problems.push(mismatch(path, 'an object', value));
        return;
      }
      for (const [key, inner] of Object.entries(value)) {
        walk(inner, schema.v, table, join(path, key), problems, depth + 1);
        if (problems.length >= MAX_PROBLEMS) return;
      }
      return;
    }

    case 'obj': {
      if (!isRecord(value)) {
        problems.push(mismatch(path, 'an object', value));
        return;
      }
      for (const [name, fieldSchema, required] of schema.f) {
        const present = name in value;
        const inner = value[name];
        if (!present || inner === undefined) {
          if (required === 1) {
            problems.push({ path: join(path, name), message: 'is required but was absent' });
          }
          continue;
        }
        walk(inner, fieldSchema, table, join(path, name), problems, depth + 1);
        if (problems.length >= MAX_PROBLEMS) return;
      }
      // Extra keys are never a problem. A server adding a field must not break a client, and this is
      // where that promise is kept.
      if (schema.a !== undefined) {
        const declared = new Set(schema.f.map(([name]) => name));
        for (const [key, inner] of Object.entries(value)) {
          if (declared.has(key)) continue;
          walk(inner, schema.a, table, join(path, key), problems, depth + 1);
          if (problems.length >= MAX_PROBLEMS) return;
        }
      }
      return;
    }

    case 'or': {
      // A union passes if any branch does. Only the *shortest* failure is reported when none do:
      // listing every branch's complaints for a three-way union is noise, and the branch that got
      // furthest is almost always the one the server meant.
      let best: ValidationProblem[] | undefined;
      for (const branch of schema.o) {
        const branchProblems: ValidationProblem[] = [];
        walk(value, branch, table, path, branchProblems, depth + 1);
        if (branchProblems.length === 0) return;
        if (best === undefined || branchProblems.length < best.length) best = branchProblems;
      }
      problems.push(...(best ?? [{ path, message: 'matched no variant of the union' }]));
      return;
    }
  }
}

function mismatch(path: string, expected: string, actual: unknown): ValidationProblem {
  return { path, message: `should be ${expected} but was ${describe(actual)}` };
}

/** A short, safe description of a value. Never interpolates the value itself — it may be a secret. */
function describe(value: unknown): string {
  if (value === null) return 'null';
  if (Array.isArray(value)) return 'an array';
  switch (typeof value) {
    case 'undefined':
      return 'absent';
    case 'string':
      return 'a string';
    case 'number':
      return Number.isInteger(value) ? 'an integer' : 'a number';
    case 'boolean':
      return 'a boolean';
    case 'object':
      return 'an object';
    default:
      return typeof value;
  }
}

function join(path: string, key: string): string {
  // A key needing quoting is rare but real; bracket form keeps the path copy-pasteable as JS.
  const safe = /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(key);
  if (path === '') return safe ? key : `[${JSON.stringify(key)}]`;
  return safe ? `${path}.${key}` : `${path}[${JSON.stringify(key)}]`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/**
 * Whether a descriptor tree contains anything worth coercing.
 *
 * Computed once per schema so the common case — a spec with no timestamps — pays nothing. Memoised on
 * the table, because the answer cannot change for a given descriptor.
 */
const coercionNeeded = new WeakMap<object, boolean>();

function needsCoercion(schema: Schema, table: SchemaTable, seen = new Set<string>()): boolean {
  switch (schema.k) {
    case 'date':
      return true;
    case 'arr':
    case 'null':
      return needsCoercion(schema.i, table, seen);
    case 'map':
      return needsCoercion(schema.v, table, seen);
    case 'or':
      return schema.o.some((branch) => needsCoercion(branch, table, seen));
    case 'obj':
      return (
        schema.f.some(([, field]) => needsCoercion(field, table, seen)) ||
        (schema.a !== undefined && needsCoercion(schema.a, table, seen))
      );
    case 'ref': {
      // A recursive type must not recurse forever here either.
      if (seen.has(schema.n)) return false;
      seen.add(schema.n);
      const target = table[schema.n];
      return target === undefined ? false : needsCoercion(target, table, seen);
    }
    default:
      return false;
  }
}

/**
 * Convert wire values into their native form, following the same descriptor tree.
 *
 * Returns the input **by reference** when nothing changed, so an unchanged response is not copied. That
 * matters: this runs on every response, and copying a large payload to change nothing would be a real
 * cost for no benefit.
 */
export function coerce(value: unknown, schema: Schema, table: SchemaTable, depth = 0): unknown {
  if (depth > MAX_DEPTH) return value;

  switch (schema.k) {
    case 'date':
      if (typeof value !== 'string') return value;
      {
        const parsed = new Date(value);
        // An unparseable timestamp is left as the string it was. Validation already reported it, and
        // handing the caller an `Invalid Date` would be strictly less useful than the original text.
        return Number.isNaN(parsed.getTime()) ? value : parsed;
      }

    case 'ref': {
      const target = table[schema.n];
      return target === undefined ? value : coerce(value, target, table, depth + 1);
    }

    case 'null':
      return value === null ? value : coerce(value, schema.i, table, depth + 1);

    case 'arr': {
      if (!Array.isArray(value)) return value;
      let changed = false;
      const mapped = value.map((item) => {
        const next = coerce(item, schema.i, table, depth + 1);
        if (next !== item) changed = true;
        return next;
      });
      return changed ? mapped : value;
    }

    case 'map': {
      if (!isRecord(value)) return value;
      let changed = false;
      const out: Record<string, unknown> = {};
      for (const [key, inner] of Object.entries(value)) {
        const next = coerce(inner, schema.v, table, depth + 1);
        if (next !== inner) changed = true;
        out[key] = next;
      }
      return changed ? out : value;
    }

    case 'obj': {
      if (!isRecord(value)) return value;
      let changed = false;
      const out: Record<string, unknown> = { ...value };
      for (const [name, fieldSchema] of schema.f) {
        if (!(name in value)) continue;
        const next = coerce(value[name], fieldSchema, table, depth + 1);
        if (next !== value[name]) {
          changed = true;
          out[name] = next;
        }
      }
      if (schema.a !== undefined) {
        const declared = new Set(schema.f.map(([name]) => name));
        for (const [key, inner] of Object.entries(value)) {
          if (declared.has(key)) continue;
          const next = coerce(inner, schema.a, table, depth + 1);
          if (next !== inner) {
            changed = true;
            out[key] = next;
          }
        }
      }
      return changed ? out : value;
    }

    case 'or': {
      // The first branch that changes something wins. A union of a timestamp and something else is
      // rare, and trying every branch and merging would be worse than taking the first that applies.
      for (const branch of schema.o) {
        const next = coerce(value, branch, table, depth + 1);
        if (next !== value) return next;
      }
      return value;
    }

    default:
      return value;
  }
}

/**
 * Apply a validation mode to a decoded body, returning it unchanged.
 *
 * Threaded through rather than validating inside the transport, because only the generated method
 * knows which schema a response has — and passing the schema down would put type knowledge in the
 * runtime, which is the boundary this design exists to keep.
 */
export function enforce<T>(
  value: T,
  schema: Schema | undefined,
  table: SchemaTable,
  operation: string,
  mode: ValidationMode,
): T {
  if (schema === undefined) return value;

  // Coercion runs even with validation off: the *types* promise a `Date`, and honouring that is not a
  // safety check a caller can reasonably decline. Turning validation off declines the *checking*, not
  // the shape of what comes back.
  const converted = (): T => {
    let needed = coercionNeeded.get(schema as object);
    if (needed === undefined) {
      needed = needsCoercion(schema, table);
      coercionNeeded.set(schema as object, needed);
    }
    return needed ? (coerce(value, schema, table) as T) : value;
  };

  if (mode === 'off') return converted();

  const problems = validate(value, schema, table);
  if (problems.length === 0) return converted();
  if (mode === 'warn') {
    // `console.warn` rather than a callback: a warning nobody wired up is a warning nobody sees, and
    // an SDK that silently drops diagnostics is worse than one that is briefly noisy.
    const error = new ResponseValidationError(operation, problems, value);
    console.warn(error.message);
    return converted();
  }
  throw new ResponseValidationError(operation, problems, value);
}

/**
 * Validate and coerce one item from a paginated response.
 *
 * Separate from {@link enforce} because a paginator has already unwrapped the envelope, and reporting
 * `[0].email` for something the caller received as a bare item would name a path they never saw. The
 * operation name still points at the method, which is what they can act on.
 */
export function enforceItem<T>(
  item: T,
  schema: Schema,
  table: SchemaTable,
  operation: string,
  mode: ValidationMode,
): T {
  return enforce(item, schema, table, operation, mode);
}
